### PR TITLE
driver-only ECC: Make PSA curves always a superset of ECP curves

### LINKED ISSUE #8014: driver-only ECC: Make PSA curves always a superset of ECP curves
There are two ways to enable support for a particular elliptic curve:
- `PSA_WANT_ECC_xxx` enables support for the curve in PSA;
- `MBEDTLS_ECP_DP_xxx` enables support for the curve in ECP and all other legacy/built-in modules.

We should make sure that the set of curves available on the PSA side is always a superset of those available on the legacy side.

I think this is currently true when `MBEDTLS_PSA_CRYPTO_CONFIG` is disabled, but it should also be true when it is enabled.

Note: this is similar to what we did for hashes.

Note: might want to wait for #7615 in order to avoid conflicts.

### PR DESCRIPTION
**Pull request by valeriosetti**
This PR ensures that the list of `ECP_DP` set is always a subset/equal to the list of supported `PSA_WANT_ECC` ones.

Resolves #8014

## PR checklist

- [ ] **changelog** question to reviewers: since we're setting requirements on the symbols the user can define in his configuration files, should we document this?
- [ ] **backport** not required since it's an enhancement
- [ ] **tests** not required

### CODE REVIEW

File: include/mbedtls/check_config.h
Comment (mpg) [inline include/mbedtls/check_config.h:71]: I think this should be just `PSA_CRYPTO_C`. We want the property "PSA curves is a superset of ECP curves" to be true in all builds (that have PSA).

**Review by mpg — COMMENTED** [review]
This looks incomplete to me. I was expecting we'd need, in some place to auto-enable the `PSA_WANT` symbols based on the `ECP_DP` ones - I believe this is already the case when `PSA_CRYPTO_CONFIG` is disabled, but not when it's enabled.

Users could be creating weird configs where they enable `PSA_CRYPTO_CONFIG`, enable a lot of curves in `mbedtls config.h`, but only a few of them in `psa/crypto_config.h`, and we don't want to have to deal with that, so we should:
1. Reject those configs - which this PR does.
2. In addition, for user convenience, but also for backwards compatibility (because those configs were never documented as being forbidden), auto-correct by enabling `PSA_WANT` macros if needed - which I think is missing.

Note: we can keep the check in `check_config.h` as a safety precaution, not really against user errors then, but against errors in our auto-enable code.

File: include/mbedtls/check_config.h
Comment (valeriosetti) [reply include/mbedtls/check_config.h:71]: I took these guards from `build_info.h` [line 105](https://github.com/Mbed-TLS/mbedtls/blob/5e678fd4d2f530e8082096b25d5ac916fb378e56/include/mbedtls/build_info.h#L105)

Question: having `MBEDTLS_PSA_CRYPTO_CONFIG` or `MBEDTLS_PSA_CRYPTO_C` isn't a superset of just `MBEDTLS_PSA_CRYPTO_C`? So the target:

>We want the property "PSA curves is a superset of ECP curves" to be true in all builds (that have PSA)

should be achieved, ins't it?

**Comment by valeriosetti** [comment]
> Users could be creating weird configs where they enable PSA_CRYPTO_CONFIG, enable a lot of curves in mbedtls config.h, but only a few of them in psa/crypto_config.h, and we don't want to have to deal with that, so we have two alternatives

Correct, I didn't think about the mixed case (i.e. when `PSA_CRYPTO_CONFIG` is enabled but curves are also defined with `ECP_DP` symbols). I'll fix that

**Comment by valeriosetti** [comment]
I had to rebase this because I needed test components from `all.sh` with accelerated ECC/FFDH without bignum for solving #8016 (which is the child issue of this PR)

**Comment by tom-daubney-arm** [comment]
I'm assuming @gilles-peskine-arm was commenting only, and not reviewing, so I have added myself as the second reviewer. Pjease correct me if that is wrong.

**Comment by gilles-peskine-arm** [comment]
@tom-daubney-arm Thanks. Even if I do review, we needed a second reviewer here since Manuel is away. I'll do at least a design review, I should be able to do a full review next week if nobody else steps up.

**Comment by tom-daubney-arm** [comment]
> @tom-daubney-arm Thanks. Even if I do review, we needed a second reviewer here since Manuel is away. I'll do at least a design review, I should be able to do a full review next week if nobody else steps up.

Great, thanks @gilles-peskine-arm. I think @tom-cosgrove-arm will take a review by the looks of it, assuming he is feeling better. In any case I would appreciate a design review from you.

**Review by tom-daubney-arm — DISMISSED** [review]
This PR looks in good shape to me. I don't wish to speak for @mpg while he is away but it looks to me like his concerns have been addressed in the follow up commits and I agree with @valeriosetti regarding the comment about which guards to use in `check_config.h`.

LGTM.

**Comment by valeriosetti** [comment]
I'm force pushing the last commit because Travis CI found had spurious errors with `arm64` build which are not related to the code.

**Review by gilles-peskine-arm — COMMENTED** [review]
Approved 4d02061447252aea293980b6b86fbb67ae9128a0 as a subset of https://github.com/Mbed-TLS/mbedtls/pull/8075#pullrequestreview-1611611971.

However, this conflicts with https://github.com/Mbed-TLS/mbedtls/pull/8159 and I'd rather not rebase that a third time in so many days. So please hold off until that is merged, then rebase/merge again.

**Comment by valeriosetti** [comment]
@gilles-peskine-arm @tom-daubney-arm I rebased this one after #8159 was merged into `development`.

Note: even though some header files were added in #8159, I didn't change my commits' text compared to the previous status, because I though this could ease your review work. I didn't change the content of the commits indeed, I just moved the changes as necessary.

File: include/mbedtls/config_adjust_legacy_from_psa.h
Comment (gilles-peskine-arm) [inline include/mbedtls/config_adjust_legacy_from_psa.h:545]: This is in the wrong file.

This is fixed in a later commit in https://github.com/Mbed-TLS/mbedtls/pull/8075.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
I don't want to merge this as is with a known problem. I'd personally prefer to close this and finish https://github.com/Mbed-TLS/mbedtls/pull/8075 which I've approved.

**Comment by mpg** [comment]
Following Gilles's suggestion, I'm closing this and we'll focus on #8075 (which includes this one) instead.

**Pull request closed without merging**
